synthtool/metadata.py

```diff
@@ -100,26 +117,61 @@ def write(outfile: str = "synth.metadata") -> None:
     log.debug(f"Wrote metadata to {outfile}.")
 
 
-def remove_obsolete_files(old_metadata):
+def _remove_obsolete_files(old_metadata):
     """Remove obsolete files from the file system.
 
     Call add_new_files() before this function or it will remove all generated
     files.
 
     Parameters:
-    old_metadata:  old metadata loaded from a call to read_or_empty().
+        old_metadata:  old metadata loaded from a call to read_or_empty().
     """
     old_files = set([new_file.path for new_file in old_metadata.new_files])
     new_files = set([new_file.path for new_file in _metadata.new_files])
     obsolete_files = old_files - new_files
-    for file_path in obsolete_files:
+    for file_path in git_ignore(obsolete_files):
         try:
             log.info(f"Removing obsolete file {file_path}...")
             os.unlink(file_path)
         except FileNotFoundError:
             pass  # Already deleted.  That's OK.
 
 
+def git_ignore(file_paths: Iterable[str]):
+    """Returns a new list of the same files, with ignored files removed."""
+    # Surprisingly, git check-ignore doesn't ignore .git directories, take those
+    # files out manually.
+    nongit_file_paths = [
+        file_path
+        for file_path in file_paths
+        if ".git" not in pathlib.Path(file_path).parts
+    ]
+    # Write the files to a temporary text file.
+    with tempfile.TemporaryFile("w+b") as f:
+        for file_path in nongit_file_paths:
+            f.write(_git_slashes(file_path).encode("utf-8"))
+            f.write("\n".encode("utf-8"))
+        # Invoke git.
+        f.seek(0)
+        git = shutil.which("git")
```

Comment:
I like this approach, you shell out to `git` rather than trying to parse the `.gitignore` file (as an alternative approach).

Reply:
Thanks.  Fewer lines of code and guaranteed to behave the same as git.